### PR TITLE
[trytond] Tracing and transaction info

### PR DESCRIPTION
**Pull request by n1ngu**
Add tracing to the Trytond integration and extract transaction information from RPC request objects.

### CODE REVIEW

**Comment by n1ngu** [comment]
Any insight on how to test this is very welcome

File: tox.ini [hunk 1]
```diff
@@ -563,15 +563,16 @@ deps =
...
 
     # Trytond
-    trytond-v4: trytond~=4.0
-    trytond-v5: trytond~=5.0
-    trytond-v6: trytond~=6.0
-    trytond-v7: trytond~=7.0
+    # Do NOT use trytond~=X.0, LTS releases are numbered as X.0.Y
+    # https://discuss.tryton.org/t/release-process/395
+    trytond-v4: trytond~=4.0.0
+    trytond-v5: trytond~=5.0.0
```
[the remark this answers is not in the document]
Comment (n1ngu) [reply tox.ini:569]: I agree! But the Sentry project has a rather conservative policy regarding extending support for old framework versions. See for example that Python 2.7 + Django 1.8 is still in the tox matrix! So I guess they'll rather keep those old versions unless there is certain incompatibility (which I might introduce with the tracing feature!).

Let's see if a maintainer shows up but, yeah, most of the Tryton community won't be interested in any <6.0 support.

**Comment by n1ngu** [comment]
@pokoli sorry to bother you. Just pushed the tracing feature commit with a fair amount of metaprogramming black magic. I am still figuring out some typing annotation details but I'd enjoy your review on that part.

It is working just fine in some production deployments but I may have missed some thoughts for multi-tenancy deployments.

imatge

**Comment by pokoli** [comment]
@n1ngu starting from 6.8 series there is a new log feature, which is used by the tryton server to log events from the user.
Maybe you will like to trace such events also in sentry. Do you think this is doable?

**Comment by n1ngu** [comment]
> @n1ngu starting from 6.8 series there is a new log feature, which is used by the tryton server to log events from the user. Maybe you will like to trace such events also in sentry. Do you think this is doable?

Because that logging is implemented with a ModelStorage, any `"ir.model.log":create` call will get its own span with the present PR! `Model.log` could be instrumented just as create, copy, delete, etc, but there is very few code there. I think it would be more interesting to create spans for `Transaction.commit` or `Transaction._store_log_records` but that can be done out of the box with
```
sentry_sdk.init(
    traces_sample_rate=1/128,
    functions_to_trace=[
        {"qualified_name": "trytond.transaction.Transaction.commit"},
        {"qualified_name": "trytond.transaction.Transaction._store_log_records"},
    ],
)
```
because, unlike pool models, Transaction can be statically instrumented. `Transaction.start` is explicitly patched in this PR because I find the whole context manager it creates deserves a span but couldn't figure out how to do that with `functions_to_trace` (I feel I am missing a more obvious solution here)

I don't know if you meant to capture log events in sentry with [`capture_message`](https://docs.sentry.io/platforms/python/usage/#capturing-messages) to send events irrespective of unhandled exceptions and tracing sample rate? That'd be certainly interesting and easily doable! I'll give the idea a try in my staging deployments while this PR settles. I feel the major challenge will be configuring this not to wipe your sentry event ingestion quota :smiling_face_with_tear:

File: tox.ini [hunk 2]
```diff
@@ -563,15 +563,16 @@ deps =
     tornado-latest: tornado
 
     # Trytond
-    trytond-v4: trytond~=4.0
-    trytond-v5: trytond~=5.0
-    trytond-v6: trytond~=6.0
-    trytond-v7: trytond~=7.0
+    # Do NOT use trytond~=X.0, LTS releases are numbered as X.0.Y
+    # https://discuss.tryton.org/t/release-process/395
+    trytond-v4: trytond~=4.0.0
```
[the remark this answers is not in the document]
Comment (n1ngu) [reply tox.ini:568]: Idem ttps://github.com/getsentry/sentry-python/issues/2653

**Review by szokeasaurusrex — CHANGES_REQUESTED** [review]
Thanks for your contribution, and thanks for your patience as we work through our backlog of community PRs.

Are you finished working on this PR? I noticed that you have added some `TODO` comments to your code. These should be addressed before we review the changes more thoroughly.

**Comment by n1ngu** [comment]
> Thanks for your contribution, and thanks for your patience as we work through our backlog of community PRs.
> 
> Are you finished working on this PR? I noticed that you have added some `TODO` comments to your code. These should be addressed before we review the changes more thoroughly.

Hi @szokeasaurusrex , thanks for looking into it.

Those TODOs are more of a wish-list or statement that I believe it'd be worth the effort to work out a better implementation. Definitely not something that prevents this PR from working rather fine as is (we could remove those comments and move on). I recall I had addressed some of them in the implementation I am deploying in production, I should give it some love and update the PR.

Yet,

1. This integration caused malfunctions under certain conditions which are still under investigation. Because of this and the lack of tests (I'd need your help to address this) I'd flag this PR as experimental. Disabling tracing should keep the integration stable but I'd not feel confident merging this as is.
2. The PR deserves some love and a rebase but we won't be able to work on it in the foreseeable mid-term due to a drift in my organization. So, feel free to close the PR if keeping it open is too noisy for you (unless the Tryton community or the Sentry organization wants to take over the work)

Cheers!

**Comment by n1ngu** [comment]
> Thanks for your contribution, and thanks for your patience as we work through our backlog of community PRs.
>
> Are you finished working on this PR? I noticed that you have added some TODO comments to your code. These should be addressed before we review the changes more thoroughly.

Hi @szokeasaurusrex , thanks for looking into it.

Those TODOs are more of a wish-list or statement that I believe it'd be worth the effort to work out a better implementation. Definitely not something that prevents this PR from working rather fine as is (we could remove those comments and move on). I recall I had addressed some of them in the implementation I am deploying in production, I should give it some love and update the PR.

Yet,

1. This integration caused malfunctions under certain conditions which are still under investigation. Because of this and the lack of tests (I'd need your help to address this) I'd flag this PR as experimental. Disabling tracing should keep the integration stable but I'd not feel confident merging this as is.
2. The PR deserves some love and a rebase but we won't be able to work on it in the foreseeable mid-term due to a drift in my organization. So, feel free to close the PR if keeping it open is too noisy for you (unless the Tryton community or the Sentry organization wants to take over the work)

Cheers!

**Comment by antonpirker** [comment]
Hey @n1ngu !

Thanks for again for your contribution. I will close this PR for now. 

Hope you can pick it up until the waves of your organizations drift have settled. 
Sentry currently also does not have the resources to pick this PR up. All the best!

**Pull request closed without merging**